v0100/srclib/dpstub.asm

```diff
@@ -29,6 +29,14 @@ __start:
         mov     [psp_seg], es
         mov     [cs_seg], cs
 
+        ; release unused DOS memory
```

Comment:
Change the comment to
"lower the stack and reduce the data segment size, releasing some memory to DOS".

Reply:
Changed to
"reduce the stack, releasing some memory to DOS".
I am really not sure why I do reduce
data. You mean bss? But its rather fixed,
and is all taken in. I don't think it can be
reduced, I only reduce the stack.

Reply:
Like I said earlier, Smaller C gives small-memory-model programs an entire 64KB segment for all of the data (const, non-const, bss) and stack. You're shrinking this whole segment.  I'm fine if you change "reduce the data segment size" to "reduce the data/stack segment size" to make it more clear.

Reply:
OK, I did what you say, but I
really don't understand. :)
My understanding is that this
memory is used _only_ by the stub.
The program itself runs in prot
mode, and uses the memory area
allocated from DPMI, not from DOS.
So if that memory is used only by
the stub, then all the data and bss
are fixed (defined as resb/resw), so
there is no way to reduce them...
I am probably missing something
obvious.

Reply:
Well, there is also an IO buffer that
is shared with the PM program, but
that one is also fixed.
So I am confused as to how could
that reduce data or bss - who allocates
them (besides the stub itself), and how?

Reply:
The stub is more or less a normal DOS .EXE. Its EXE header tells DOS how many 16-byte paragraphs are needed at a minimum and how many paragraphs it may have at a maximum. The EXE header (and these min/max sizes in it) only cover the stub itself.

The concatenated 32-bit a.out portion is simply ignored by DOS.

The minimum size (from EXE header) covers the code, data and stack of the stub. When DOS executes it, it adds the PSP size to it and allocates that much memory (unless there's more and the max size (from EXE header) asks for more). 

The EXE header is naturally created by the linker.

The stub persists in memory until the 32-bit a.out part terminates. The stub with its I/O buffer are below 1MB (the buffer, fbuf, is in stub's .bss). The buffer is needed so the DPMI code can use DOS and BIOS functions operating on memory buffers, which can't be above 1MB (not talking about HMA right now).

Reply:
Yes, this part is quite clear.
But I can't shrink the buffer,
because it is defined with
`fbuf        resb FBUFSZ`
The only way to shrink the
buffer is to reduce `FBUFSZ`.
I put `end_of_memory` after
that buffer, and any other bss
data.
This means, I can't shrink data
or bss. I can only shrink stack!
To shrink bss or data, I'd have
to put `end_of_memory` before
some of its symbols, eg before
`fbuf`. Then I indeed shrink bss,
though nothing will work.
But if I put `end_of_memory`
after bss, then how can I shrink
bss?

Reply:
Maybe you mean I shrink data
_segment_, not the data section?
And speaking of sections, I only
shrink the stack?

Reply:
Precisely, the stack is at the end of the 64KB segment and you take away some memory from that end, from the stack and therefore the entire segment.

Reply:
OK, not sure why comment needs
to say that rather than just reducing
stack, but fine. :)